src/shared/loop-util.c

```diff
@@ -109,6 +109,44 @@ static int device_has_block_children(int nr) {
         return !!sd_device_enumerator_get_device_first(e);
 }
 
+static int attach_empty_file(int loop) {
+        _cleanup_close_ int fd = -1;
+
+        /* So here's the thing: on various kernels (5.8 at least) loop block devices might enter a state
+         * where they are detached but nonetheless have partitions, when used heavily. Accessing these
+         * partitions results in immediatey IO errors. There's no pretty way to get rid of them
+         * again. Neither LOOP_CLR_FD nor LOOP_CTL_REMOVE suffice (see above). What does work is to
+         * reassociate them with a new fd however. This is what we do here hence: we associate the devices
+         * with an empty file (i.e. an image that definitely has no partitions). We then immediately clear it
+         * again. This suffices to make the partitions go away. Ugly but appears to work. */
```

Comment:
so, back when i added this code i wasn't aware of BLKPG_DEL_PARTITION. I wonder if a loop around that wouldn't have the same effect but be less ugly?

i.e. remove all partitions in a loop, until none are left?